util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,238 @@
+
+
+def get_token(auth_url: str, api_key: str) -> str:
+    """
+    This function gets the token from the database using the api key
+    :param auth_url: Authentication url to use Atlas data API
+    :param api_key: API key to access the database
+    :return: Token to access the database
+    """
+
+    data = {"key": api_key}
+    data = json.dumps(data).encode("utf-8")
+
+    req = request.Request(
+        auth_url, data=data, headers={"content-type": "application/json"}
+    )
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get token: {e}")
+
+    token = json.loads(response.read().decode("utf-8"))["access_token"]
+    return token
+
+
+def get_all_resources(
+    url: str, data_source: str, collection: str, database: str, token: str
+) -> List:
+    """
+    This function gets all the JSON objects for resources from the database
+    :param url: Database url to use Atlas data API
+    :param data_source: Data source name for the mongoDB database
+    :param collection: Collection name for the mongoDB database
+    :param database: Database name for the mongoDB database
+    :param token: Token to access the database
+
+    :return: List of JSON objects for resources
+    """
+
+    url = f"{url}/action/find"
+    data = {
+        "dataSource": data_source,
+        "collection": collection,
+        "database": database,
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get resources: {e}")
+
+    return json.loads(response.read().decode("utf-8"))["documents"]
+
+
+def save_resources_to_file(resources: List, output: pathlib.Path):
+    """
+    This function saves all the JSON objects for resources to a file.
+    :param resources: List of JSON objects for resources
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("resources.json")
+
+    with open(path, "w") as f:
+        json.dump(resources, f, indent=4)
+
+
+def progress_hook(t):
+    last_b = [0]
+
+    def update_to(b=1, bsize=1, tsize=None):
+        if tsize not in (None, -1):
+            t.total = tsize
+        displayed = t.update((b - last_b[0]) * bsize)
+        last_b[0] = b
+        return displayed
+
+    return update_to
+
+
+def download_resources(output: pathlib.Path):
+    """
+    This function downloads the resources which have a url field and
+    updates the url field to point to the local download of the resource.
+    :param output: Output directory absolute path
+    """
+
+    path = output.joinpath("offline_resources")
+    if not path.exists():
+        path.mkdir()
+
+    with open(output.joinpath("resources.json")) as f:
+        resources = json.load(f)
```

Comment:
First of all, can we split this into two functions `download_resources` and `update_urls`? This feels to me to be two seperate tasks and decoupling them would make this code a little easier to follow.

In addition, can we take the reading and writing of the JSON files out. Can we just have the signature of this function be:

```py
def download_resources(resources: List[Dict], offline_resources_dir: Path) -> None:
```

it takes the resources input and download them all to the specified output directory. To make things easier when creating the URLs after this you can state in the docstring you download all resources to files/directories named "[id]-[version]".